File: IT/Oggetti/Esercizi/C++/esercizio18.cpp
/*
file esercizio18.cpp
compilare con: g++ -std=c++11 esercizio18.cpp
*/

#include <iostream>

using std::cout;
using std::endl;
using std::string;
using std::ostream;
using std::to_string;

class Engine {
	private:
		int cc;

	public:
		Engine(int);
		string str() const;
};

Engine::Engine(int cc) {
	this->cc = cc;
}

string Engine::str() const {
	return "motore di " + to_string(this->cc) + " centimetri cubi";
}

class Vehicle {
	private:
		Engine *engine;

	public:
		Vehicle(Engine*);
		~Vehicle();
		Engine* getEngine() const;
		void setEngine(Engine*);
		virtual string str() const = 0;

	friend ostream& operator<<(ostream &os, const Vehicle&);
};

ostream& operator<<(ostream &os, const Vehicle& v) {
	os << v.str();
	return os;
}

Vehicle::Vehicle(Engine *e) {
	this->setEngine(e);
}

Vehicle::~Vehicle() {
	delete this->engine;
}

Engine* Vehicle::getEngine() const {
	return this->engine;
}

void Vehicle::setEngine(Engine* engine) {
	this->engine = engine;
}

class Car : public Vehicle {
	private:
		string model;
		string plate;
		string vin;

	public:
		Car(string, string, string, Engine *);

		void setModel(string);
		string getModel();
		void setPlate(string);
		string getPlate();
		void setVin(string);
		string getVin();
		string str() const override;
};

Car::Car(string model, string plate, string vin, Engine *engine) : Vehicle(engine) {
	setModel(model);
	setPlate(plate);
	setVin(vin);
}

string Car::getModel() {
	return this->model;
}

void Car::setModel(string model) {
	this->model = model;
}

string Car::getPlate() {
	return this->plate;
}

void Car::setPlate(string plate) {
	this->plate = plate;
}

string Car::getVin() {
	return this->vin;
}

void Car::setVin(string vin) {
	this->vin = vin;
}

string Car::str() const {
	return this->model + ", targa \"" + this->plate + "\", n. telaio \"" + this->vin + "\", " + this->getEngine()->str();
}

class Truck : public Vehicle {
	public:
		Truck(Engine*);
		string str() const override;
};

Truck::Truck(Engine* e) : Vehicle(e) {}

string Truck::str() const {
	return "Camion con " + this->getEngine()->str();
}

class Tractor : public Vehicle {
	public:
		Tractor(Engine*);
		string str() const override;
};

Tractor::Tractor(Engine* e) : Vehicle(e) {}

string Tractor::str() const {
	return "Trattore con " + this->getEngine()->str();
}

int main()
{
	Car fiat500("Fiat 500", "AS 123 DF", "1G8AN12F04Z242828", new Engine(1400));
	Car classeA("Mercedes Classe A", "CA 128 TD", "2FZHAZBD21AH02482", new Engine(1600));
	Car alfaGiulia("Alfa Romeo Giulia", "EE 119 CA", "2G1115SL7F9174905", new Engine(2000));

	cout << fiat500 << endl;
	cout << classeA << endl;
	cout << alfaGiulia << endl;

	Truck truck(new Engine(15000));
	Tractor tractor(new Engine(5800));

	cout << truck << endl;
	cout << tractor << endl;
}

File: IT/Oggetti/Esercizi/Python/esercizio18.py
# file esercizio18.py

from __future__ import print_function

class Engine(object):
	def __init__(self, cc):
		self.__cc = cc

	def __str__(self):
		return "motore di {} centimetri cubi".format(self.__cc)

class Vehicle(object):
	def __init__(self, engine):
		self.__engine = engine

	def getEngine(self):
		return self.__engine

	def setEngine(self, e):
		self.__engine = e

	engine = property(getEngine, setEngine)

class Car(Vehicle):
	def __init__(self, model, vin, plate, engine):
		Vehicle.__init__(self, engine)
		self.model = model
		self.vin = vin
		self.plate = plate

	def getModel(self):
		return self.__model

	def setModel(self, model):
		self.__model = model

	def getVin(self):
		return self.__vin

	def setVin(self, vin):
		self.__vin = vin

	def getPlate(self):
		return self.__plate

	def setPlate(self, plate):
		self.__plate = plate

	model = property(getModel, setModel)
	vin = property(getVin, setVin)
	plate = property(getPlate, setPlate)

	def __str__(self):
		return "{}, targa \"{}\", n. telaio \"{}\", {}".format(self.model, self.plate, self.vin, self.engine)

class Truck(Vehicle):
	def __str__(self):
		return "Camion con {}".format(self.engine)

class Tractor(Vehicle):
	def __str__(self):
		return "Trattore con {}".format(self.engine)

fiat500 = Car(model="Fiat 500", vin="1G8AN12F04Z242828", plate="AS 123 DF", engine=Engine(cc=1400))
classeA = Car(model="Mercedes Classe A", vin="2FZHAZBD21AH02482", plate="CA 128 TD", engine=Engine(cc=1600))
alfaGiulia = Car(model="Alfa Romeo Giulia", vin="2G1115SL7F9174905", plate="EE 119 CA", engine=Engine(cc=2000))

print(fiat500)
print(classeA)
print(alfaGiulia)

truck = Truck(engine=Engine(cc=15000))
tractor = Tractor(engine=Engine(cc=5800))

print(truck)
print(tractor)

